### PR TITLE
Add Notification System API

### PR DESCRIPTION
**Pull request by shivani-sangwan**
Adding Notification and UserNotification Association database layer.

## How to test the changes?
- Tested the relationships and functioning using the unit tests for the above-mentioned database tables. 

## License
- [x] I agree to license these contributions under [Galaxy's current license](https://github.com/galaxyproject/galaxy/blob/dev/LICENSE.txt).
- [x] I agree to allow the Galaxy committers to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT). If this condition is an issue, uncheck and just let us know why with an e-mail to galaxy-committers@lists.galaxyproject.org.

### CODE REVIEW

File: lib/galaxy/managers/notification.py [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+    ObjectNotFound,
+)
+from galaxy.managers.base import decode_id
+from galaxy.managers.context import ProvidesAppContext
+from galaxy.schema.fields import EncodedDatabaseIdField
+from galaxy.structured_app import MinimalManagerApp
+from galaxy.web import url_for
+
+
+class NotificationManager:
```
Comment (mvdbeek) [inline lib/galaxy/managers/notification.py:15]: This is looking good, but can you remove all id encoding and decoding from the Manager ? I don't think we need this at all for a modern API. This should all just work with integers.

File: lib/galaxy/managers/notification.py [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+        """
+        message_text = payload.get("Message", None)
+        if message_text is None:
+            raise ObjectAttributeMissingException("Missing required message")
+
+        notification = model.Notification(message_text=message_text)
+        trans.sa_session.add(notification)
+        encoded_user_ids = payload.get("user_ids", [])
+        users = self._get_users_by_encoded_ids(trans, encoded_user_ids)
+        trans.app.security_agent.set_entity_notification_associations(users=[users], notification=notification)
```
Comment (mvdbeek) [inline lib/galaxy/managers/notification.py]: Why did you move this into the security_agent module ? I think that's business logic that belongs in the manager

File: lib/galaxy/managers/notification.py [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+from galaxy.schema.fields import EncodedDatabaseIdField
+from galaxy.structured_app import MinimalManagerApp
+from galaxy.web import url_for
+
+
+class NotificationManager:
+    """Interface/service object shared by controllers for interacting with notifications."""
+
+    def __init__(self, app: MinimalManagerApp) -> None:
+        self._app = app
```
Comment (mvdbeek) [inline lib/galaxy/managers/notification.py]: You should only need the database session here. If a Manager doesn't need app (a [god-object](https://en.wikipedia.org/wiki/God_object), we want to remove its usage whenever possible) it shouldn't receive it.

File: lib/galaxy/managers/notification.py [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+            raise ObjectAttributeMissingException("Missing required message")
+
+        notification = model.Notification(message_text=message_text)
+        trans.sa_session.add(notification)
+        encoded_user_ids = payload.get("user_ids", [])
+        users = self._get_users_by_encoded_ids(trans, encoded_user_ids)
+        trans.app.security_agent.set_entity_notification_associations(users=[users], notification=notification)
+        trans.sa_session.flush()
+        item = group.to_dict(view="element", value_mapper={"id": trans.security.encode_id})
+        item["url"] = url_for("notification", id=encoded_id)
```
Comment (mvdbeek) [inline lib/galaxy/managers/notification.py]: We also want to eliminate usage of `url_for`, is there a specific use you have in mind for the url ?

File: lib/galaxy/managers/notification.py [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+        return [item]
```
Comment (mvdbeek) [inline lib/galaxy/managers/notification.py]: Any reason you're return a list here ?

File: test/unit/data/model/test_notification.py [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+
+def create_notification(sa_session: galaxy_scoped_session):
+    notification = Notification()
+    sa_session.add(notification)
+    sa_session.flush()
+    assert notification.id is not None
+    return notification
+
+def create_user(sa_session: galaxy_scoped_session):
+    random_number = random.randint(0, 1000)
```
Comment (mvdbeek) [inline test/unit/data/model/test_notification.py:32]: I think the next step here is to use the Manager in these tests ... you're essentially testing that you're using sqlalchemy correctly currently. That's OK, but ideally the unit tests should test our own classes and methods.

File: lib/galaxy/managers/notification.py (hunk 4 above)
Comment (shivani-sangwan) [reply lib/galaxy/managers/notification.py]: I was thinking that notifications will be posted to a specific /notification/ page for the user.

File: lib/galaxy/managers/notification.py (hunk 5 above)
Comment (shivani-sangwan) [reply lib/galaxy/managers/notification.py]: Not a specific reason, just thought if we will have group/role objects then we might need it.

File: lib/galaxy/managers/notification.py (hunk 2 above)
Comment (shivani-sangwan) [reply lib/galaxy/managers/notification.py]: Thought to follow the similar pipeline, as done for other associations like groups.py for UserGroupAssociation.

File: lib/galaxy/managers/notification.py (hunk 3 above)
Comment (shivani-sangwan) [reply lib/galaxy/managers/notification.py]: Should I use MinimalManagerTool instead, it contains a galaxy scoped session.

File: lib/galaxy/managers/notification.py (hunk 4 above)
Comment (mvdbeek) [reply lib/galaxy/managers/notification.py]: Yes, but why do you need the URL ?

File: lib/galaxy/managers/notification.py (hunk 5 above)
Comment (mvdbeek) [reply lib/galaxy/managers/notification.py]: Yes, but why wrap it in a list ?

File: lib/galaxy/managers/notification.py (hunk 3 above)
Comment (mvdbeek) [reply lib/galaxy/managers/notification.py]: No, you don't need app here at all.

File: lib/galaxy/managers/notification.py [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+    """Interface/service object shared by controllers for interacting with notifications."""
+
+
+    def index(self, trans: ProvidesAppContext):
+        """
+        Displays a collection (list) of notifications.
+        """
+        rval = []
+        for notification in trans.sa_session.query(model.Notification).filter(model.Notification.deleted == false()):
+            item = notification.to_dict(value_mapper={"id": notification.id})
```
Comment (mvdbeek) [inline lib/galaxy/managers/notification.py]:
```suggestion
```

File: lib/galaxy/managers/notification.py [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+            rval.append(item)
```
Comment (mvdbeek) [inline lib/galaxy/managers/notification.py]:
```suggestion
            rval.append(notification)
```

File: lib/galaxy/managers/notification.py [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+        return rval
+
+    def create(self, trans: ProvidesAppContext, payload: Dict[str, Any]):
```
Comment (mvdbeek) [inline lib/galaxy/managers/notification.py]:
```suggestion
    def create(self, trans: ProvidesAppContext, message: str):
```

File: lib/galaxy/managers/notification.py [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+        """
+        Creates a new notification.
+        """
+        message_text = payload.get("Message", None)
+        if message_text is None:
+            raise ObjectAttributeMissingException("Missing required message")
+
+        notification = model.Notification(message_text=message_text)
```
Comment (mvdbeek) [inline lib/galaxy/managers/notification.py]:
```suggestion
        notification = model.Notification(message_text=message)
```

File: lib/galaxy/managers/notification.py [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+        trans.sa_session.add(notification)
+        trans.sa_session.flush()
+        encoded_user_ids = payload.get("user_ids", [])
+        users = self._get_users_by_encoded_ids(trans, encoded_user_ids)
+        self._associate_user_notification(users=[users], notification=notification)
```
Comment (mvdbeek) [inline lib/galaxy/managers/notification.py]:
```suggestion
```

File: lib/galaxy/managers/notification.py [hunk 12]
```diff
@@ -0,0 +1,94 @@
...
+        return notification
+
+    def show(self, trans: ProvidesAppContext, user_id):
+        """
+        Displays information about a notification.
+        """
+        notification = self._get_notification(trans, user_id)
+        item = notification.to_dict(view="element", value_mapper={"id": notification.id})
+        return item
```
Comment (mvdbeek) [inline lib/galaxy/managers/notification.py]:
```suggestion
    def get(self, notification_id) -> model.Notification:
        """
        Get a single notification
        """
        return = self._get_notification(trans, notification_id)
```

File: lib/galaxy/managers/notification.py [hunk 13]
```diff
@@ -0,0 +1,94 @@
...
+from galaxy.managers.base import decode_id
+from galaxy.managers.context import ProvidesAppContext
+from galaxy.schema.fields import EncodedDatabaseIdField
+from galaxy.structured_app import MinimalToolApp
+from galaxy.web import url_for
+
+
+class NotificationManager:
+    """Interface/service object shared by controllers for interacting with notifications."""
+
```
Comment (mvdbeek) [inline lib/galaxy/managers/notification.py:17]:
```suggestion
    """Interface/service object shared by controllers for interacting with notifications."""

    def __init__(self, sa_session: galaxy_scoped_session):
        self.sa_session = sa_session

```

File: lib/galaxy/managers/notification.py [hunk 14]
```diff
@@ -0,0 +1,94 @@
...
+
+    def index(self, trans: ProvidesAppContext):
+        """
+        Displays a collection (list) of notifications.
+        """
+        rval = []
+        for notification in trans.sa_session.query(model.Notification).filter(model.Notification.deleted == false()):
```
Comment (mvdbeek) [inline lib/galaxy/managers/notification.py]:
```suggestion
        for notification in self.sa_session.query(model.Notification).filter(model.Notification.deleted == false()):
```

File: lib/galaxy/managers/notification.py [hunk 15]
```diff
@@ -0,0 +1,94 @@
...
+from galaxy.schema.fields import EncodedDatabaseIdField
+from galaxy.structured_app import MinimalToolApp
+from galaxy.web import url_for
+
+
+class NotificationManager:
+    """Interface/service object shared by controllers for interacting with notifications."""
+
+
+    def index(self, trans: ProvidesAppContext):
```
Comment (mvdbeek) [inline lib/galaxy/managers/notification.py]:
```suggestion
    def index(self, limit: int = 10, offset: int = 0, user: Optional[model.User] = None):
```

Comment (mvdbeek) [reply lib/galaxy/managers/notification.py]: I'll leave the implementation of filtering for users, limit and offset as an exercise for you to implement

File: lib/galaxy/managers/notification.py [hunk 16]
```diff
@@ -0,0 +1,94 @@
...
+        """
+        notification = self._get_notification(trans, encoded_id)
+        message_text = payload.get("Message", None)
+        if message_text:
+            notification.message_text = message_text
+            trans.sa_session.add(notification)
+        trans.sa_session.flush()
+        encoded_user_ids = payload.get("user_ids", [])
+        users = self._get_users_by_encoded_ids(trans, encoded_user_ids)
+        self._associate_user_notification(users=[users], notification=notification)
```
Comment (mvdbeek) [inline lib/galaxy/managers/notification.py]:
```suggestion
    def update(self, notification: model.Notification, message: str):
        """
        Modifies a notification.
        """
        notification.message_text = message_text
        trans.sa_session.add(notification)
        trans.sa_session.flush()
        return notification
```

File: lib/galaxy/managers/notification.py [hunk 17]
```diff
@@ -0,0 +1,94 @@
...
+
+
+
+    def _get_notification(self, trans: ProvidesAppContext, user_id) -> model.Notification:
+        notification = trans.sa_session.query(model.Notification).get(user_id)
+        if notification is None:
+            raise ObjectNotFound(f"Notification with id {user_id} was not found.")
+        return notification
```
Comment (mvdbeek) [inline lib/galaxy/managers/notification.py]:
```suggestion
    def _get_notification(self, user_id: int) -> model.Notification:
        notification = self.sa_session.query(model.Notification).get(user_id)
        if notification is None:
            raise ObjectNotFound(f"Notification with id {user_id} was not found.")
        return notification
```

File: lib/galaxy/managers/notification.py [hunk 18]
```diff
@@ -0,0 +1,94 @@
...
+
+    def _get_users_by_encoded_ids(
+        self, trans: ProvidesAppContext, encoded_user_ids: List[EncodedDatabaseIdField]
+    ) -> List[model.User]:
+        decoded_user_ids = self._decode_ids(encoded_user_ids)
+        users = trans.sa_session.query(model.User).filter(model.User.table.c.id.in_(decoded_user_ids)).all()
+        return users
+
```
Comment (mvdbeek) [inline lib/galaxy/managers/notification.py]:
```suggestion
```

That's what the user manager does, that doesn't need to go here.

File: lib/galaxy/managers/notification.py [hunk 19]
```diff
@@ -0,0 +1,94 @@
...
+    def _associate_user_notification(self, user, notification):
+        for user in users:
+            assoc = model.UserNotificationAssociation(user, notification)
+            trans.sa_session.add(assoc)
+            trans.sa_session.flush()
```
Comment (mvdbeek) [inline lib/galaxy/managers/notification.py]:
```suggestion
    def _associate_user_notification(self, users: List[model.user], notification: model.Notification):
        for user in users:
            assoc = model.UserNotificationAssociation(user, notification)
            trans.sa_session.add(assoc)
        trans.sa_session.flush()
```

File: lib/galaxy/managers/notification.py [hunk 20]
```diff
@@ -0,0 +1,94 @@
...
+from galaxy.exceptions import (
+    Conflict,
+    ObjectAttributeMissingException,
+    ObjectNotFound,
+)
+from galaxy.managers.base import decode_id
+from galaxy.managers.context import ProvidesAppContext
+from galaxy.schema.fields import EncodedDatabaseIdField
+from galaxy.structured_app import MinimalToolApp
+from galaxy.web import url_for
```
Comment (mvdbeek) [inline lib/galaxy/managers/notification.py]:
```suggestion
```

File: lib/galaxy/managers/notification.py [hunk 21]
```diff
@@ -0,0 +1,61 @@
...
+from galaxy.exceptions import (
+    ObjectNotFound,
+)
+from galaxy.model.scoped_session import galaxy_scoped_session
+
+class NotificationManager:
+    """Interface/service object shared by controllers for interacting with notifications."""
+
+    def __init__(self, sa_session: galaxy_scoped_session):
+        self.sa_session = galaxy_scoped_session
```
Comment (mvdbeek) [inline lib/galaxy/managers/notification.py]:
```suggestion
        self.sa_session = sa_session
```
this is why your tests are failing

File: lib/galaxy/model/__init__.py [hunk 22]
```diff
@@ -2409,6 +2409,42 @@ def __init__(self, user, group):
...
+class UserNotificationAssociation(Base, RepresentById):
+    __tablename__ = "user_notification_association"
+
+    id = Column(Integer, primary_key=True)
+    user_id = Column(Integer, ForeignKey("galaxy_user.id"), index=True)
+    notification_id = Column(Integer, ForeignKey("notification_push.id"), index=True)
+    create_time = Column(DateTime, default=now)
+    update_time = Column(DateTime, default=now, onupdate=now)
+    user = relationship("User", backref="all_notifications")
+    notification = relationship("Notification", backref="user_notification_associations")
```
Comment (jdavcs) [inline lib/galaxy/model/__init__.py]: Please use `back_populates` instead of `backref` (you'll also need to add the opposite relationship to the `Notification` class. (see #13165 for context)

File: lib/galaxy/model/__init__.py [hunk 23]
```diff
@@ -8127,7 +8163,6 @@ def _serialize(self, id_encoder, serialization_options):
         request_input_collection_attrs["order_index"] = self.workflow_step.order_index
         return request_input_collection_attrs
 
-
```
Comment (jdavcs) [inline lib/galaxy/model/__init__.py]: 2 spaces between class definitions is correct

File: lib/galaxy/model/__init__.py [hunk 24]
```diff
@@ -8176,7 +8211,6 @@ def _serialize(self, id_encoder, serialization_options):
         output_dataset_attrs["workflow_output"] = self.workflow_output.serialize(id_encoder, serialization_options)
         return output_dataset_attrs
 
-
```
Comment (jdavcs) [inline lib/galaxy/model/__init__.py]: same as above

**Comment by jdavcs** [comment]
Instead of the `test/unit/data/model/test_notifications.py` module, please add the appropriate tests to `test/unit/data/model/mapping/test_gxy_model_mapping.py`: that module is precisely for what you were testing. There's documentation at the top of that module. So, you'd add a new `Test[classname]` class for each of your model classes, testing their table, columns, and relationship mappings. And you'd need to add a line to the `TestUser`  class too.

File: test/unit/data/model/test_notification.py (hunk 6 above)
Comment (jdavcs) [reply test/unit/data/model/test_notification.py:32]: we have infrastructure for that - see my separate comment on testing the mapping.

**Comment by mvdbeek** [comment]
@jdavcs the tests will be testing the managers, I don't think they belong in the mapping tests, those are different types of tests.

**Comment by jdavcs** [comment]
> @jdavcs the tests will be testing the managers, I don't think they belong in the mapping tests, those are different types of tests.

Oh, I see, I was probably looking at the lone assert stmt in the create_user function (and the comment about the tests testing sqlalchemy usage). So yes, then just please add the mapping tests for the new tables (*not instead* of the manager tests).

File: lib/galaxy/schema/schema.py [hunk 25]
```diff
@@ -3043,6 +3057,14 @@ class Config:
         extra = Extra.allow  # Allow any other extra fields
 
 
+class CreateNotificationPayload(NotificationBase):
+    notification: Optional[str] = NotificationContentField
+    
+    class Config:
+        use_enum_values = True  # When using .dict()
+        extra = Extra.allow  # Allow any other extra fields
```
Comment (mvdbeek) [inline lib/galaxy/schema/schema.py]: why did you add this?

File: lib/galaxy/managers/notification.py [hunk 26]
```diff
@@ -0,0 +1,75 @@
...
+        for notification in (
+            self.sa_session.query(model.Notification)
+            .filter(model.Notification.deleted == false())
+            .limit(limit)
+            .offset(limit + offset)
+        ):
+            rval.append(notification)
+        return rval
+
+    def create(self, trans, payload):
```
Comment (mvdbeek) [inline lib/galaxy/managers/notification.py]: You do not need trans here, you can use self.sa_session. You should also add the type annotation for payload here.

File: lib/galaxy/schema/schema.py [hunk 27]
```diff
@@ -3003,6 +3004,19 @@ class PageSummaryBase(BaseModel):
...
 
 
+NotificationContentField: Optional[str] = Field(
+    default="",
+    title="Notification",
+    description="Raw text contents of the notification message.",
+)
+
+class NotificationBase(BaseModel):
+    message: str = Field(
```
Comment (mvdbeek) [inline lib/galaxy/schema/schema.py]: This should be message_text to be consistent with the database model.
You also need a response model with the properties you want to return. That can be the same model for now.

File: lib/galaxy/schema/schema.py [hunk 28]
```diff
@@ -3043,6 +3057,14 @@ class Config:
         extra = Extra.allow  # Allow any other extra fields
 
 
+class CreateNotificationPayload(NotificationBase):
+    notification: Optional[str] = NotificationContentField
```
Comment (mvdbeek) [inline lib/galaxy/schema/schema.py]: I think you're overcomplicating things here. All you need in the CreateNotificationPayload is the message_text property, i don't think this requires separately defined fields, just do define it here.

File: lib/galaxy/schema/schema.py [hunk 29]
```diff
@@ -3043,6 +3057,14 @@ class Config:
...
+    
+    class Config:
+        use_enum_values = True  # When using .dict()
```
Comment (mvdbeek) [inline lib/galaxy/schema/schema.py]: You're not using any enums

File: lib/galaxy/schema/schema.py [hunk 30]
```diff
@@ -46,6 +46,7 @@
 
 USER_MODEL_CLASS_NAME = "User"
 GROUP_MODEL_CLASS_NAME = "Group"
+NOTIFICATION_MODEL_CLASS_NAME = "Notification"
```
Comment (mvdbeek) [inline lib/galaxy/schema/schema.py]: you're not using this

File: lib/galaxy/webapps/galaxy/api/notifications.py [hunk 31]
```diff
@@ -0,0 +1,51 @@
...
+        return self.manager.index(limit=5)
+
+    @router.get(
+        "/api/notifications",
+        summary="Displays information about a notification.",
+    )
+    def show(self, notification_id):
+        return self.manager.show(notification_id)
+
+    @router.post("/api/notifications", summary="Create a notificaton message")
```
Comment (mvdbeek) [inline lib/galaxy/webapps/galaxy/api/notifications.py]: You need to annotate a response model here, you can't just return the database model here. This is true for all the methods here.

File: lib/galaxy/webapps/galaxy/api/notifications.py [hunk 32]
```diff
@@ -0,0 +1,65 @@
...
+        return self.manager.index(limit=5)
+
+    @router.get(
+        "/api/notifications",
+        summary="Displays information about a notification.",
+    )
+    def show(
+        self,
+        trans: ProvidesAppContext = DependsOnTrans,
+        notification_id: NotificationIdField = NotificationIdParam,
```
Comment (mvdbeek) [inline lib/galaxy/webapps/galaxy/api/notifications.py]:
```suggestion
        notification_id: NotificationIdParam,
        trans: ProvidesAppContext = DependsOnTrans,
```

File: lib/galaxy/webapps/galaxy/api/notifications.py [hunk 33]
```diff
@@ -0,0 +1,65 @@
...
+    ) -> NotificationModel:
+        return self.manager.show(trans, notification_id)
+
+    @router.post("/api/notifications", summary="Create a notificaton message")
+    def create(
+        self,
+        trans: ProvidesAppContext = DependsOnTrans,
+        message_text: NotificationMessageField = NotificationMessageParam,
```
Comment (mvdbeek) [inline lib/galaxy/webapps/galaxy/api/notifications.py:59]:
```suggestion
        message_text: NotificationMessageParam,
```

Comment (mvdbeek) [inline lib/galaxy/webapps/galaxy/api/notifications.py]:
```suggestion
        message_text: NotificationMessageParam,
        trans: ProvidesAppContext = DependsOnTrans,
```

File: lib/galaxy/webapps/galaxy/api/notifications.py [hunk 34]
```diff
@@ -0,0 +1,65 @@
...
+    Router,
+    DependsOnTrans
+)
+
+log = logging.getLogger(__name__)
+
+router = Router(tags=["notifications"])
+
+NotificationIdParam: NotificationIdField = Path(..., title="ID", description="Notification ID of the notification")
+
```
Comment (mvdbeek) [inline lib/galaxy/webapps/galaxy/api/notifications.py]:
```suggestion
```

File: lib/galaxy/webapps/galaxy/api/notifications.py [hunk 35]
```diff
@@ -0,0 +1,65 @@
...
+NotificationMessageParam: NotificationMessageField = Path(..., title="Message", description="Message of the notification")
+
```
Comment (mvdbeek) [inline lib/galaxy/webapps/galaxy/api/notifications.py:27]:
```suggestion


```

File: lib/galaxy_test/api/test_notifications.py [hunk 36]
```diff
@@ -0,0 +1,50 @@
...
+class NotificationsApiTestCase(ApiTestCase):
+    def setUp(self):
+        super().setUp()
+        self.dataset_populator = DatasetPopulator(self.galaxy_interactor)
+
+    def test_index(self):
+        payload = {"message_text": "Test Message"}
+        notification_response = self._post("notifications", payload, json=True)
+        notification_response.raise_for_status()
+        payload = {"limit": 5, "offset": 0}
```
Comment (mvdbeek) [inline lib/galaxy_test/api/test_notifications.py]: The types of your notification dictionary are changing, that's why you're getting:
```
lib/galaxy_test/api/test_notifications.py:14: error: Dict entry 0 has
incompatible type "str": "int"; expected "str": "str"  [dict-item]
            payload = {"limit": 5, "offset": 0}
```

File: lib/galaxy_test/api/test_notifications.py [hunk 37]
```diff
@@ -0,0 +1,50 @@
...
+        notification_response = self._get("notifications", payload)
```
Comment (mvdbeek) [inline lib/galaxy_test/api/test_notifications.py]:
```suggestion
        create_notification_payload = {"message_text": "Test Message"}
        notification_response = self._post("notifications", create_notification_payload, json=True)
        notification_response.raise_for_status()
        notification_list_parameters = {"limit": 5, "offset": 0}
        notification_response = self._get("notifications", notification_list_parameters)
```

File: client/src/components/Notification/List.vue [hunk 38]
```diff
@@ -0,0 +1,59 @@
+<template>
+    <div class="notifications-card">
+        <b-alert variant="error" v-if="errorMessage" show > </b-alert>
+        <loading-span v-if="loading" message="Loading notifications" />
+        <ul v-else id="notifications">
+            <li v-for="notifiation in notifications" :key="notifications" :data-extension-target="notifications">
+                <span
```
Comment (dannon) [inline client/src/components/Notification/List.vue]:
```suggestion
                <span>
```
This span open tag is missing its right arm.

File: client/src/components/Notification/List.vue [hunk 39]
```diff
@@ -0,0 +1,59 @@
...
+                    <div class="ui success message">
+                        <i class="close icon"></i>
+                        <div class="header">
+                           {{ notifiation.message_text }}
+                        </div>
```
Comment (dannon) [inline client/src/components/Notification/List.vue]:
```suggestion
                        </div>
                    </div>
```

Missing a closing div here.

**Comment by nsoranzo** [comment]
Thanks a lot for all the work on this @shivani-sangwan ! Work on the Notification System, based on your contributions, is continuing on https://github.com/galaxyproject/galaxy/pull/15663 .
